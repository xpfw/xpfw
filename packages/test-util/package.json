{
  "name": "@xpfw/test-util",
  "version": "1.0.0-beta.7",
  "main": "dist/index",
  "typings": "dist/index",
  "license": "MIT",
  "dependencies": {
    "@feathersjs/authentication": "^2.1.15",
    "@feathersjs/authentication-jwt": "^2.0.9",
    "@feathersjs/authentication-local": "^1.2.9",
    "@feathersjs/express": "^1.3.1",
    "@feathersjs/feathers": "^3.3.1",
    "@feathersjs/socketio": "^3.2.9",
    "@xpfw/validate": "0.1.1",
    "feathers-memory": "^3.0.1",
    "feathers-mongodb": "^4.0.2",
    "lodash": "^4.17.11",
    "moment": "^2.23.0",
    "mongodb": "^3.1.10"
  },
  "scripts": {
    "build": "./node_modules/.bin/rollup -c",
    "dev": "./node_modules/.bin/rollup -c -w",
    "test": "./node_modules/.bin/jest --ci",
    "update": "ncu -a",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@types/feathersjs__authentication": "^2.1.2",
    "@types/feathersjs__authentication-jwt": "^1.0.3",
    "@types/feathersjs__authentication-local": "^1.0.3",
    "@types/feathersjs__errors": "^3.2.1",
    "@types/feathersjs__express": "^1.1.4",
    "@types/feathersjs__feathers": "^3.1.1",
    "@types/feathersjs__socketio": "^3.0.3",
    "@types/jest": "^23.3.12",
    "@types/lodash": "^4.14.119",
    "@types/mongodb": "^3.1.18",
    "@types/node": "^10.12.18",
    "jest": "^23.6.0",
    "jest-cli": "^23.6.0",
    "mockdate": "^2.0.2",
    "rollup": "^1.1.0",
    "rollup-plugin-typescript2": "^0.18.1",
    "rollup-plugin-uglify": "^6.0.0",
    "ts-jest": "=23.10.5",
    "typescript": "^3.2.2",
    "uglify-es": "^3.3.10"
  },
  "jest": {
    "globals": {
      "ts-jest": {
        "diagnostics": false
      }
    },
    "transform": {
      ".(ts|tsx)": "ts-jest"
    },
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|tsx|js)$",
    "testPathIgnorePatterns": [
      "\\.snap$",
      "<rootDir>/node_modules/",
      "<rootDir>/lib",
      "<rootDir>/dist"
    ],
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js"
    ],
    "testURL": "http://localhost/"
  }
}
