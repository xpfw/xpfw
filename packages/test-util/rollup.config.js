// rollup.config.js
import typescript from "rollup-plugin-typescript2"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript()
  ],
  external: [
    "lodash",
    "moment",
    "body-parser",
    "@feathersjs/feathers",
    "@feathersjs/authentication",
    "feathers-hooks",
    "feathers-memory",
    "feathers-mongodb",
    "@feathersjs/express/rest",
    "@feathersjs/errors/handler",
    "@feathersjs/express",
    "@feathersjs/socketio",
    "@feathersjs/authentication-local",
    "@feathersjs/authentication-jwt",
    "mongodb",
    "@xpfw/data",
    "http"
  ]
};