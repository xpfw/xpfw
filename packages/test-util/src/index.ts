import clearService from "./testUtil/clearService"
import ensureNoObjectId from "./testUtil/ensureNoObjectId"
import getRandomApp from "./testUtil/getRandomApp"
import matchHookError from "./testUtil/matchHookError"
import stripId from "./testUtil/stripId"

export {
  getRandomApp, stripId, ensureNoObjectId, matchHookError, clearService
}
