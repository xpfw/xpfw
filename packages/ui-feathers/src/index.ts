import FeathersClient from "./feathers"

export default FeathersClient
export {
  FeathersClient
}
