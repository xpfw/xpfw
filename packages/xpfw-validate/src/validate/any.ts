import { get, intersection, isFunction, isNil, isNumber, isObject, isString, keys, set } from "lodash"
import { ErrorType, FieldType, Method } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateArray from "./array"
import validateBool from "./bool"
import validateDate from "./date"
import validateLocation from "./location"
import validateNumber from "./number"
import validateObject from "./object"
import validateQueryObject, { possibleQueryOperators } from "./queryObject"
import validateSelect from "./select"
import validateText from "./text"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateAny: ValidateFunction = (value, field, params) => {
  let errorResolve: any = null
  if (get(params, "method") === Method.Find && isObject(value)) {
    if (intersection(keys(value), possibleQueryOperators).length > 0) {
      const form: any = params.form
      return validateQueryObject(value, form, params)
    }
  }
  if (isObject(field) && !isNil(field.validate)) {
    if (isNumber(field.type)) {
      switch (field.type) {
        case FieldType.Text:
        case FieldType.Password:
        case FieldType.RelationshipSingle: {
          errorResolve = validateText(value, field, params)
          break
        }
        case FieldType.Number:
        case FieldType.Slider: {
          errorResolve = validateNumber(value, field, params)
          break
        }
        case FieldType.Boolean: {
          errorResolve = validateBool(value, field, params)
          break
        }
        case FieldType.Date: {
          errorResolve = validateDate(value, field, params)
          break
        }
        case FieldType.Select: {
          errorResolve = validateSelect(value, field, params)
          break
        }
        case FieldType.Object: {
          errorResolve = validateObject(value, field, params)
          break
        }
        case FieldType.Array:
        case FieldType.RelationshipMulti: {
          if (field.type === FieldType.RelationshipMulti) {
            set(field, "validate.type", FieldType.Text)
          }
          errorResolve = validateArray(value, field, params)
          break
        }
        case FieldType.Location: {
          errorResolve = validateLocation(value, field, params)
          break
        }
        default: {
          errorResolve = null
          break
        }
      }
    } else if (isString(field.type)) {
      if (isObject(params) && isObject(params.registry)) {
        const validator = params.registry.findValidator(field.type)
        if (isFunction(validator)) {
          errorResolve = validator(value, field, params)
        }
      }
    }
    if (isNil(errorResolve)) {
      console.log("erroirrseolve is null for", field)
      errorResolve = Promise.reject(false)
    }
  } else {
    errorResolve = Promise.resolve(value)
  }
  if (isObject(field) && isObject(field.validate)) {
    const validate: any = field.validate
    if (isFunction(validate.custom)) {
      let catched = false
      const errs: any = {mapTo: field.mapTo, errors: []}
      return errorResolve.catch((e: any) => {
        catched = true
        errs.errors = e
        return validate.custom(value, field, params)
      }).then(() => {
        if (!catched) {
          return validate.custom(value, field, params)
        }
        return Promise.reject(errs)
      })
    }
  }
  return errorResolve
}

export default validateAny
