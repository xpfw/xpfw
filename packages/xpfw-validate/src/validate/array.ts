import { cloneDeep, each, get, isArray, isNil, isNumber, isString } from "lodash"
import { ErrorType, FieldType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateAny from "./any"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateArray: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validArray = isArray(value)
  if (checkShouldBeRequired(field, params) && !validArray) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (validArray) {
    const min = get(field, "validate.min")
    if (isNumber(min)) {
      if (value.length < min) {
        errors.push({
          type: ErrorType.tooLow
        })
      }
    }
    const max = get(field, "validate.max")
    if (isNumber(max)) {
      if (value.length > max) {
        errors.push({
          type: ErrorType.tooHigh
        })
      }
    }
  }
  if (errors.length > 0) {
    return Promise.reject(errors)
  }
  const promises: Array<Promise<any>> = []
  if (!isNil(field.validate)) {
    const subValidate: any = field.validate
    if (isString(subValidate.type) || isNumber(subValidate.type)) {
      each(value, (date) => {
        const tmpFieldOptions = cloneDeep(field)
        tmpFieldOptions.type = subValidate.type
        if (tmpFieldOptions.type === FieldType.Array || tmpFieldOptions.type === FieldType.Object) {
          tmpFieldOptions.validate = subValidate.validate
        }
        promises.push(validateAny(date, tmpFieldOptions, params))
      })
    }
  }
  return Promise.all(promises).then((results) => {
    return value
  })
}

export default validateArray
