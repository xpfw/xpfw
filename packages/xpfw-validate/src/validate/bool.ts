import { isBoolean, isNil } from "lodash"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateBool: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const isValidBoolean = isBoolean(value)
  if (checkShouldBeRequired(field, params) && (!isValidBoolean || !value)) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidBoolean) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateBool
