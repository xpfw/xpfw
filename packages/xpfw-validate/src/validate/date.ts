import { isDate, isNil, isString } from "lodash"
import * as moment from "moment"
import { ErrorType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import checkShouldBeRequired from "./util/checkShouldBeRequired"
const m: any = moment
const validateDate: ValidateFunction = (value, field, params) => {
  if (isString(value) && m(value).isValid()) {
    value = m(value).toDate()
  }
  const errors: IErrorBase[] = []
  const isValidDate = isDate(value)
  if (checkShouldBeRequired(field, params) && !isValidDate) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (!isNil(value) && !isValidDate) {
    errors.push({
      type: ErrorType.invalidType
    })
  }
  if (errors.length > 0) {
    return Promise.reject({mapTo: field.mapTo, errors})
  }
  return Promise.resolve(value)
}

export default validateDate
