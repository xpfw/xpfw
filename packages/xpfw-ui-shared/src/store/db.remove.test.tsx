import { FormStore } from "@xpfw/form-shared"
import { getRandomApp } from "@xpfw/test-util"
import { FeathersClient } from "@xpfw/ui-feathers"
import { TestDefs } from "@xpfw/validate"
import "isomorphic-fetch"
import { get } from "lodash"
import { matchStoreState } from "resub-persist"
import BackendClient from "../client"
import DbStore from "./db"

BackendClient.client = FeathersClient

test("DbStore Remove Test", async () => {
  const s: any = TestDefs.FormNumberAndRequiredText.collection
  const appRef = await getRandomApp(s, false, BackendClient.client, true)
  matchStoreState(DbStore, "before anything")

  FormStore.setValue(TestDefs.RequiredTextField.mapTo, "myText")
  FormStore.setValue(TestDefs.NumberField.mapTo, 420)
  const createdObject = await DbStore.create(TestDefs.FormNumberAndRequiredText)
  expect(createdObject).toMatchSnapshot("createresult")
  FormStore.setValue(TestDefs.RequiredTextField.mapTo, "changedText")
  FormStore.setValue(TestDefs.NumberField.mapTo, 123)
  const removeRes = await DbStore.remove(get(createdObject, "result.id"), TestDefs.FormNumberAndRequiredText.collection)
  expect(removeRes).toEqual(createdObject)
  expect(removeRes).toMatchSnapshot("removeresult")
  matchStoreState(DbStore, "After calling remove")
  const getResult = await DbStore.getFromServer(get(createdObject, "result.id"),
    TestDefs.FormNumberAndRequiredText.collection)
  expect(get(getResult, "error")).not.toBeNull()
  expect(getResult).not.toEqual(removeRes)
  expect(getResult).not.toEqual(createdObject)
  expect(removeRes).toEqual(DbStore.getRemoveState(get(createdObject, "result.id")))

  TestDefs.FormNumberAndRequiredText.collection = "broken"
  const res = await DbStore.remove(get(createdObject, "result.id"), TestDefs.FormNumberAndRequiredText.collection)
  expect(get(res, "error")).not.toBeNull()
  await appRef.cleanUp()
}, 10000)
