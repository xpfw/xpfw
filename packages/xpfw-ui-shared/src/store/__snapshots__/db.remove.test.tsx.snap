// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbStore Remove Test: After calling remove 1`] = `
Object {
  "createState": Object {
    "formModel": Object {
      "result": Object {
        "id": 0,
        "myNum": 420,
        "myString": "myText",
      },
    },
  },
  "getState": Object {
    "simpleTestCol": Object {},
  },
  "updateState": Object {},
}
`;

exports[`DbStore Remove Test: before anything 1`] = `
Object {
  "createState": Object {},
  "getState": Object {},
  "updateState": Object {},
}
`;

exports[`DbStore Remove Test: createresult 1`] = `
Object {
  "result": Object {
    "id": 0,
    "myNum": 420,
    "myString": "myText",
  },
}
`;

exports[`DbStore Remove Test: removeresult 1`] = `
Object {
  "result": Object {
    "id": 0,
    "myNum": 420,
    "myString": "myText",
  },
}
`;
